python/ray/serve/_private/router.py

```diff
@@ -385,42 +387,69 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
         """Shim for compatibility with the existing round robin scheduler."""
         return self.update_replicas([ActorReplicaWrapper(r) for r in running_replicas])
 
-    def _get_candidate_replica_ids(
+    def _get_candidate_replica_ids_for_multiplexed_model_id(
         self,
-        blacklist_replica_ids: Set[str],
-        request_metadata: Optional[RequestMetadata] = None,
+        model_id: str,
+        get_from_all_replicas: bool = False,
     ) -> Set[str]:
-        """Get candidates from the current replica set excluding the blacklist.
+        """Get multiplexed model candidates from the current replica.
+
+        By default, we will only choose from replicas that have the requested
+        multiplexed model id, if not matched, the function will return an empty set.
+
+        If get_from_all_replicas is True, we will choose from all replicas,
+        and we will choose all replicas with the least number of multiplexed model
+        ids.
 
-        If a model ID is present in request_metadata, any replicas that have it are
-        prioritized.
         """
-        if (
-            request_metadata is not None
-            and request_metadata.multiplexed_model_id
-            in self._multiplexed_model_id_to_replica_ids
-        ):
-            candidates = self._multiplexed_model_id_to_replica_ids[
-                request_metadata.multiplexed_model_id
-            ].difference(blacklist_replica_ids)
-            if len(candidates) > 0:
-                return candidates
 
-        return self._replica_id_set.difference(blacklist_replica_ids)
+        candidates = set()
+
+        if not get_from_all_replicas:
+            if model_id in self._multiplexed_model_id_to_replica_ids:
+                candidates = self._multiplexed_model_id_to_replica_ids[model_id]
+                if len(candidates) > 0:
+                    return candidates
+            return candidates
```

Comment:
I think my previous comment wasn't addressed. As written this means that we will always wait the `RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S` on the first request when there are no replicas that have a given model. Is there any way to avoid this fixed static cost?

Reply:
Updated the logic. Currently "first" request will fallback immediately, and all other requests for same model id will wait for matching till  RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S.